meteor/server/api/playout/adlib.ts

```diff
@@ -411,6 +411,48 @@ export namespace ServerPlayoutAdLibAPI {
 		})
 	}
 
+	export function innerFindLastScriptedPieceOnLayer(
+		cache: CacheForPlayout,
+		sourceLayerId: string,
+		customQuery?: MongoQuery<Piece>
+	) {
+		const span = profiler.startSpan('innerFindLastScriptedPieceOnLayer')
+
+		const playlist = cache.Playlist.doc
+		const rundownIds = getRundownIDsFromCache(cache)
+
+		if (!playlist.currentPartInstanceId) {
+			return
+		}
+
+		const currentPartInstance = cache.PartInstances.findOne({ _id: playlist.currentPartInstanceId })
+
+		if (!currentPartInstance) {
+			return
+		}
+
+		const query = {
+			...customQuery,
+			startRundownId: { $in: rundownIds },
+			sourceLayerId,
+		}
+
+		if (span) span.end()
+
+		const pieces = Pieces.find(query).fetch()
+
+		const part = Parts.findOne(
+			{ _id: { $in: pieces.map((p) => p.startPartId) }, _rank: { $lte: currentPartInstance.part._rank } },
+			{ sort: { _rank: -1 } }
+		)
+
+		if (!part) {
+			return
+		}
+
+		return pieces.find((p) => p.startPartId === part._id)
```

Comment:
Im not keen on this. It wont be deterministic (maybe not even completely consistent) if there are multiple matching pieces. If you filter for all the pieces in the part, and sort those by start time that will produce better results.

It might be appropriate to run them through `processAndPrunePieceInstanceTimings` to do the start time sorting and also collision handling to match how they would be played

Reply:
We can't use `processAndPrunePieceInstanceTimings ` because we're working with pieces, so I have sorted the pieces acording to `enable.start`

Reply:
True, but you could wrap the pieces to pieceInstances to run them through the function, and then unwrap the result. The ui does that if it is working on a part rather than partinstance

You would need to ensure you load in every piece in the part for this to be correct. maybe not its infinites though?